off by 1 error in os.copy (200USD Bounty)

https://scastie.scala-lang.org/5htwvdwbShivvYgwZZgqUg

```
Caused by: os.PathError$AbsolutePathOutsideRoot$: The path created has enough ..s that it would start outside the root directory
	at os.PathError$AbsolutePathOutsideRoot$.&lt;clinit&gt;(Path.scala:220)
	at os.Path.$div(Path.scala:462)
	at os.copy$.apply(FileOps.scala:179)
```
copying to a subpath works, copying to the root folder does not.

**Haoyi**: *To incentivize contribution, I'm putting a 200USD bounty on resolving this ticket. This is payable via bank transfer, and at my discretion in case of ambiguity.*

--- a/os/src/FileOps.scala
+++ b/os/src/FileOps.scala
@@ -84,7 +84,7 @@ object move {
       atomicMove: Boolean = false,
       createFolders: Boolean = false
   ): Unit = {
-    if (createFolders) makeDir.all(to / up)
+    if (createFolders && to.segmentCount != 0) makeDir.all(to / up)
     val opts1 =
       if (replaceExisting) Array[CopyOption](StandardCopyOption.REPLACE_EXISTING)
       else Array[CopyOption]()
@@ -176,7 +176,7 @@ object copy {
       createFolders: Boolean = false,
       mergeFolders: Boolean = false
   ): Unit = {
-    if (createFolders) makeDir.all(to / up)
+    if (createFolders && to.segmentCount != 0) makeDir.all(to / up)
     val opts1 =
       if (followLinks) Array[CopyOption]()
       else Array[CopyOption](LinkOption.NOFOLLOW_LINKS)

--- a/os/src/FileOps.scala
+++ b/os/src/FileOps.scala
@@ -68,7 +68,7 @@ object move {
       def isDefinedAt(x: Path) = partialFunction.isDefinedAt(x)
       def apply(from: Path) = {
         val dest = partialFunction(from)
-        makeDir.all(dest / up)
+        if (dest.segmentCount != 0) makeDir.all(dest / up)
         os.move(from, dest, replaceExisting, atomicMove, createFolders)
       }
     }
@@ -149,7 +149,7 @@ object copy {
       def isDefinedAt(x: Path) = partialFunction.isDefinedAt(x)
       def apply(from: Path) = {
         val dest = partialFunction(from)
-        makeDir.all(dest / up)
+        if (dest.segmentCount != 0) makeDir.all(dest / up)
         os.copy(
           from,
           dest,

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -214,7 +214,8 @@ object PathTestsCustomFilesystem extends TestSuite {
             os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
           } match {
             // This is expected. We just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
-            case Failure(e @(_: IllegalArgumentException | _: FileAlreadyExistsException)) => e.getMessage
+            case Failure(e @ (_: IllegalArgumentException | _: FileAlreadyExistsException)) =>
+              e.getMessage
           }
         }
       }

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -206,13 +206,11 @@ object PathTestsCustomFilesystem extends TestSuite {
           assert(os.exists(root / "file.txt"))
         }
       }
-      test("moveMatchingToRootDirectory") {
+      test("moveMatchingToRootDirectoryShouldFail") {
         withCustomFs { fileSystem =>
-          try {
+          // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
+          intercept[FileAlreadyExistsException] {
             os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
-          } catch {
-            case e: PathError.AbsolutePathOutsideRoot.type => throw e
-            case NonFatal(_) => ()
           }
         }
       }

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -7,6 +7,7 @@ import java.nio.file.FileSystems
 import java.net.URI
 import java.nio.file.FileSystem
 import java.nio.file.Paths
+import scala.util.control.NonFatal
 
 object PathTestsCustomFilesystem extends TestSuite {
 
@@ -173,6 +174,28 @@ object PathTestsCustomFilesystem extends TestSuite {
           os.copy(p / "file.txt", p / "file2.txt")
           assert(os.read(p / "file2.txt") == "Hello World")
           assert(os.exists(p / "file.txt"))
+        }
+      }
+      test("copyAndMergeToRootDirectoryWithCreateFolders") {
+        withCustomFs { fileSystem =>
+          val root = os.root("/", fileSystem)
+          val file = root / "test" / "dir" / "file.txt"
+          os.write(file, "Hello World")
+          os.copy(root / "test" / "dir", root, createFolders = true, mergeFolders = true)
+          assert(os.read(root / "file.txt") == "Hello World")
+          assert(os.exists(root / "file.txt"))
+        }
+      }
+      test("moveToRootDirectoryWithCreateFolders") {
+        withCustomFs { fileSystem =>
+          // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
+          val root = os.root("/", fileSystem)
+          try {
+            os.move(root / "test" / "dir", root, createFolders = true)
+          } catch {
+            case e: PathError.AbsolutePathOutsideRoot.type => throw e
+            case NonFatal(_) => ()
+          }
         }
       }
       test("remove") {

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -213,8 +213,8 @@ object PathTestsCustomFilesystem extends TestSuite {
           Try {
             os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
           } match {
-            // This is expected. We just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
-            case Failure(e @ (_: IllegalArgumentException | _: FileAlreadyExistsException)) =>
+            case Failure(e @ (_: IllegalArgumentException | _: FileAlreadyExistsException))
+                if !e.isInstanceOf[PathError.AbsolutePathOutsideRoot.type] =>
               e.getMessage
           }
         }

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -198,6 +198,28 @@ object PathTestsCustomFilesystem extends TestSuite {
           }
         }
       }
+      test("copyMatchingAndMergeToRootDirectory") {
+        withCustomFs { fileSystem =>
+          val root = os.root("/", fileSystem)
+          val file = root / "test" / "dir" / "file.txt"
+          os.write(file, "Hello World")
+          os.list(root / "test").collect(os.copy.matching(mergeFolders = true) {
+            case p / "test" / _ => p
+          })
+          assert(os.read(root / "file.txt") == "Hello World")
+          assert(os.exists(root / "file.txt"))
+        }
+      }
+      test("moveMatchingToRootDirectory") {
+        withCustomFs { fileSystem =>
+          try {
+            os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
+          } catch {
+            case e: PathError.AbsolutePathOutsideRoot.type => throw e
+            case NonFatal(_) => ()
+          }
+        }
+      }
       test("remove") {
         withCustomFs { fileSystem =>
           val p = os.root("/", fileSystem) / "test" / "dir"

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -206,11 +206,13 @@ object PathTestsCustomFilesystem extends TestSuite {
           assert(os.exists(root / "file.txt"))
         }
       }
-      test("moveMatchingToRootDirectoryShouldFail") {
+      test("moveMatchingToRootDirectory") {
         withCustomFs { fileSystem =>
-          // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
-          intercept[FileAlreadyExistsException] {
+          try {
             os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
+          } catch {
+            case e: PathError.AbsolutePathOutsideRoot.type => throw e
+            case NonFatal(_) => ()
           }
         }
       }

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -2,11 +2,10 @@ package test.os
 
 import utest._
 import os._
+
 import java.util.HashMap
-import java.nio.file.FileSystems
+import java.nio.file.{FileAlreadyExistsException, FileSystem, FileSystems}
 import java.net.URI
-import java.nio.file.FileSystem
-import java.nio.file.Paths
 import scala.util.control.NonFatal
 
 object PathTestsCustomFilesystem extends TestSuite {
@@ -186,15 +185,12 @@ object PathTestsCustomFilesystem extends TestSuite {
           assert(os.exists(root / "file.txt"))
         }
       }
-      test("moveToRootDirectoryWithCreateFolders") {
+      test("moveToRootDirectoryWithCreateFoldersSouldFail") {
         withCustomFs { fileSystem =>
-          // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
           val root = os.root("/", fileSystem)
-          try {
+          // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
+          intercept[FileAlreadyExistsException] {
             os.move(root / "test" / "dir", root, createFolders = true)
-          } catch {
-            case e: PathError.AbsolutePathOutsideRoot.type => throw e
-            case NonFatal(_) => ()
           }
         }
       }

--- a/os/test/src-jvm/PathTestsCustomFilesystem.scala
+++ b/os/test/src-jvm/PathTestsCustomFilesystem.scala
@@ -6,6 +6,7 @@ import os._
 import java.util.HashMap
 import java.nio.file.{FileAlreadyExistsException, FileSystem, FileSystems}
 import java.net.URI
+import scala.util.{Failure, Try}
 import scala.util.control.NonFatal
 
 object PathTestsCustomFilesystem extends TestSuite {
@@ -15,7 +16,7 @@ object PathTestsCustomFilesystem extends TestSuite {
     path.toUri()
   }
 
-  def withCustomFs(f: FileSystem => Unit, fsUri: URI = customFsUri()): Unit = {
+  def withCustomFs[T](f: FileSystem => T, fsUri: URI = customFsUri()): T = {
     val uri = new URI("jar", fsUri.toString(), null);
     val env = new HashMap[String, String]();
     env.put("create", "true");
@@ -185,7 +186,7 @@ object PathTestsCustomFilesystem extends TestSuite {
           assert(os.exists(root / "file.txt"))
         }
       }
-      test("moveToRootDirectoryWithCreateFoldersSouldFail") {
+      test("failMoveToRootDirectoryWithCreateFolders") {
         withCustomFs { fileSystem =>
           val root = os.root("/", fileSystem)
           // This should fail. Just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
@@ -206,13 +207,14 @@ object PathTestsCustomFilesystem extends TestSuite {
           assert(os.exists(root / "file.txt"))
         }
       }
-      test("moveMatchingToRootDirectory") {
+      test("failMoveMatchingToRootDirectory") {
         withCustomFs { fileSystem =>
-          try {
+          // can't use a `intercept`, see https://github.com/com-lihaoyi/os-lib/pull/267#issuecomment-2116131445
+          Try {
             os.list(os.root("/", fileSystem)).collect(os.move.matching { case p / "test" => p })
-          } catch {
-            case e: PathError.AbsolutePathOutsideRoot.type => throw e
-            case NonFatal(_) => ()
+          } match {
+            // This is expected. We just test that it doesn't throw PathError.AbsolutePathOutsideRoot.
+            case Failure(e @(_: IllegalArgumentException | _: FileAlreadyExistsException)) => e.getMessage
           }
         }
       }

